api/wasm/v0/cpp/proxy_wasm_externs.h

```diff
@@ -0,0 +1,126 @@
+//  of this ABI e.g. the C++ SDK provides a proxy_wasm_api.h implementation of the API on top of
+//  this ABI.
+//
+
+// Non-stream calls.
+
+/**
+ * Called when the VM starts for each plugin.
+ * @param root_context_id is an identifier for one or more related plugins and their corresponding
+ * proxy_on_configure(), proxy_on_done() and proxy_on_delete() calls. It is also provided during
+ * creation of stream context for those plugins.
+ * @param vm_configuration_size is the size of any configuration available via
+ * proxy_get_configuration during the lifetime of this call.
+ * @return non-zero on success and zero on failure (e.g. bad configuration).
+ */
+extern "C" uint32_t proxy_on_vm_start(uint32_t root_context_id, uint32_t vm_configuration_size);
+
+/**
+ * Can be called to validate a configuration (e.g. from bootstrap or xDS) both before
+ * proxy_on_start() to verify the VM configuration or after proxy_on_start() to verify a plugin
+ * configuration.
+ * @param root_context_id is a unique identifier for the configuration verification context.
+ * @param configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ * @return non-zero on success and zero on failure (i.e. bad configuration).
+ */
+extern "C" uint32_t proxy_validate_configuration(uint32_t root_context_id,
+                                                 uint32_t configuration_size);
+/**
+ * Called when a plugin loads or when plugin configuration changes dynamically.
+ * @param root_context_id is an identifier for one or more related plugins.
+ * @param plugin_configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ * @return non-zero on success and zero on failure (e.g. bad configuration).
+ */
+extern "C" uint32_t proxy_on_configure(uint32_t root_context_id,
+                                       uint32_t plugin_configuration_size);
+
+// Stream calls.
+
+/**
+ * Called when a request, stream or other ephemeral context is created.
+ * @param context_id is an identifier the ephemeral context.
+ * @param configuration_size is the size of any configuration available via
+ * proxy_get_configuration().
+ */
+extern "C" void proxy_on_context_create(uint32_t context_id, uint32_t root_context_id);
+
+// Stream and Non-stream calls.
+
+/**
+ * For stream contexts, called when the stream has completed. Note: if applicable proxy_on_log() is
+ * called after proxy_on_done() and before proxy_on_delete(). For root contexts, proxy_on_done() is
+ * called when the VM is going to shutdown.
+ * @param context_id is an identifier the context.
+ * @return non-zero to indicate that this context is done. Stream contexts must return non-zero.
+ * Root contexts may return zero to defer the VM shutdown and the proxy_on_delete call until after a
+ * future proxy_done() call by the root context.
+ */
+extern "C" uint32_t proxy_on_done(uint32_t context_id);
```

Comment:
can this return an enum of actions to take vs. a 0 / non-zero return?

Reply:
It is only 0 and non-zero.  Perhaps in the future we could differentiate non-zero values.  Do you want to specify this as 0 or 1 and leave the other values as reserved?

Reply:
What I mean is I think it would be a lot more clear to have it return an enum, even if right now the enum only has 2 values.

Reply:
Added enums for all returns values of on_XXX functions.  PTAL